File: app/src/main/java/com/huanchengfly/tieba/post/fragments/ThreadStoreFragment.java
package com.huanchengfly.tieba.post.fragments;


import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.material.snackbar.Snackbar;
import com.huanchengfly.tieba.api.TiebaApi;
import com.huanchengfly.tieba.api.models.CommonResponse;
import com.huanchengfly.tieba.api.models.ThreadStoreBean;
import com.huanchengfly.tieba.post.R;
import com.huanchengfly.tieba.post.adapters.ThreadStoreAdapter;
import com.huanchengfly.tieba.post.components.MyLinearLayoutManager;
import com.huanchengfly.tieba.post.components.dividers.RecycleViewDivider;
import com.huanchengfly.tieba.post.models.database.Account;
import com.huanchengfly.tieba.post.utils.AccountUtil;
import com.huanchengfly.tieba.post.utils.NavigationHelper;
import com.huanchengfly.tieba.post.utils.SharedPreferencesUtil;
import com.huanchengfly.tieba.post.utils.ThemeUtil;
import com.huanchengfly.tieba.post.utils.Util;

import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ThreadStoreFragment extends BaseFragment {

    public NavigationHelper navigationHelper;
    private RecyclerView recyclerView;
    private ThreadStoreAdapter threadStoreAdapter;
    private SwipeRefreshLayout refreshLayout;
    private int page = 0;
    private boolean hasMore = true;
    private String tbs;

    public ThreadStoreFragment() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        navigationHelper = NavigationHelper.newInstance(getAttachContext());
        Account account = AccountUtil.getLoginInfo(getAttachContext());
        if (account != null) tbs = account.getTbs();
    }

    @Override
    int getLayoutId() {
        return R.layout.fragment_thread_store;
    }

    @NotNull
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View contentView = super.onCreateView(inflater, container, savedInstanceState);
        refreshLayout = contentView.findViewById(R.id.thread_store_refresh_layout);
        refreshLayout.setOnRefreshListener(this::refresh);
        ThemeUtil.setThemeForSwipeRefreshLayout(refreshLayout);
        recyclerView = contentView.findViewById(R.id.thread_store_recycler_view);
        recyclerView.setLayoutManager(new MyLinearLayoutManager(getAttachContext()));
        recyclerView.addItemDecoration(new RecycleViewDivider(getAttachContext(), LinearLayoutManager.VERTICAL, R.drawable.drawable_divider_1dp));
        threadStoreAdapter = new ThreadStoreAdapter(getAttachContext());
        threadStoreAdapter.setLoadingView(R.layout.layout_footer_loading);
        threadStoreAdapter.setLoadEndView(R.layout.layout_footer_loadend);
        threadStoreAdapter.setLoadFailedView(R.layout.layout_footer_load_failed);
        threadStoreAdapter.setOnLoadMoreListener(this::loadMore);
        threadStoreAdapter.setOnItemClickListener((viewHolder, threadStoreInfo, i) -> {
            Map<String, String> map = new HashMap<>();
            map.put("tid", threadStoreInfo.getThreadId());
            map.put("pid", threadStoreInfo.getMarkPid());
            map.put("seeLz", SharedPreferencesUtil.get(getAttachContext(), SharedPreferencesUtil.SP_SETTINGS).getBoolean("collect_thread_see_lz", true) ? "1" : "0");
            map.put("from", "collect");
            map.put("max_pid", threadStoreInfo.getMaxPid());
            navigationHelper.navigationByData(NavigationHelper.ACTION_THREAD, map);
        });
        ItemTouchHelper mItemTouchHelper = new ItemTouchHelper(new ItemTouchHelper.Callback() {
            @Override
            public void clearView(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
                super.clearView(recyclerView, viewHolder);
                viewHolder.itemView.setBackgroundColor(Color.TRANSPARENT);
            }

            @Override
            public void onSelectedChanged(RecyclerView.ViewHolder viewHolder, int actionState) {
                super.onSelectedChanged(viewHolder, actionState);
                if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
                    viewHolder.itemView.setBackgroundColor(Util.getColorByAttr(getAttachContext(), R.attr.colorControlHighlight, R.color.transparent));
                }
            }

            @Override
            public int getMovementFlags(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
                int dragFlags = 0, swiped = ItemTouchHelper.RIGHT | ItemTouchHelper.LEFT;
                return makeMovementFlags(dragFlags, swiped);
            }

            @Override
            public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
                return true;
            }

            @Override
            public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
                int position = viewHolder.getAdapterPosition();
                ThreadStoreBean.ThreadStoreInfo threadStoreInfo = threadStoreAdapter.getData(position);
                threadStoreAdapter.remove(position);
                Util.createSnackbar(recyclerView, R.string.toast_deleted, Snackbar.LENGTH_LONG)
                        .addCallback(new Snackbar.Callback() {
                            @Override
                            public void onDismissed(Snackbar snackbar, int event) {
                                if (event != DISMISS_EVENT_ACTION) {
                                    TiebaApi.getInstance()
                                            .removeStore(threadStoreInfo.getThreadId(), tbs)
                                            .enqueue(new Callback<CommonResponse>() {
                                                @Override
                                                public void onResponse(@NotNull Call<CommonResponse> call, @NotNull Response<CommonResponse> response) {
                                                }

                                                @Override
                                                public void onFailure(@NotNull Call<CommonResponse> call, @NotNull Throwable t) {
                                                    Toast.makeText(getAttachContext(), getAttachContext().getString(R.string.toast_delete_error, t.getMessage()), Toast.LENGTH_SHORT).show();
                                                    threadStoreAdapter.insert(threadStoreInfo, position);
                                                }
                                            });
                                }
                            }
                        }).setAction(R.string.button_undo, mView -> threadStoreAdapter.insert(threadStoreInfo, position)).show();
            }
        });
        mItemTouchHelper.attachToRecyclerView(recyclerView);
        recyclerView.setAdapter(threadStoreAdapter);
        return contentView;
    }

    @Override
    protected void onFragmentFirstVisible() {
        refresh();
    }

    private void refresh() {
        refreshLayout.setRefreshing(true);
        page = 0;
        TiebaApi.getInstance()
                .threadStore(page, 20)
                .enqueue(new Callback<ThreadStoreBean>() {
                    @Override
                    public void onResponse(@NotNull Call<ThreadStoreBean> call, @NotNull Response<ThreadStoreBean> response) {
                        ThreadStoreBean data = response.body();
                        refreshLayout.setRefreshing(false);
                        List<ThreadStoreBean.ThreadStoreInfo> storeInfoList = data.getStoreThread();
                        if (storeInfoList == null) {
                            return;
                        }
                        threadStoreAdapter.reset();
                        threadStoreAdapter.setNewData(storeInfoList);
                        hasMore = storeInfoList.size() > 0;
                        if (!hasMore) {
                            threadStoreAdapter.loadEnd();
                        }
                    }

                    @Override
                    public void onFailure(@NotNull Call<ThreadStoreBean> call, @NotNull Throwable t) {
                        refreshLayout.setRefreshing(false);
                    }
                });
    }

    private void loadMore(boolean isReload) {
        if (!isReload) {
            page += 1;
        }
        if (!hasMore) {
            return;
        }
        TiebaApi.getInstance()
                .threadStore(page, 20)
                .enqueue(new Callback<ThreadStoreBean>() {
                    @Override
                    public void onResponse(@NotNull Call<ThreadStoreBean> call, @NotNull Response<ThreadStoreBean> response) {
                        ThreadStoreBean data = response.body();
                        List<ThreadStoreBean.ThreadStoreInfo> storeInfoList = data.getStoreThread();
                        if (storeInfoList == null) {
                            return;
                        }
                        threadStoreAdapter.setLoadMoreData(storeInfoList);
                        hasMore = storeInfoList.size() > 0;
                        if (!hasMore) {
                            threadStoreAdapter.loadEnd();
                        }
                    }

                    @Override
                    public void onFailure(@NotNull Call<ThreadStoreBean> call, @NotNull Throwable t) {
                        threadStoreAdapter.loadFailed();
                        Toast.makeText(getAttachContext(), t.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/fragments/ThreadStoreFragment.kt
package com.huanchengfly.tieba.post.fragments

import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import com.google.android.material.snackbar.Snackbar
import com.huanchengfly.tieba.post.R
import com.huanchengfly.tieba.post.adapters.ThreadStoreAdapter
import com.huanchengfly.tieba.post.adapters.base.OnItemClickListener
import com.huanchengfly.tieba.post.api.TiebaApi.getInstance
import com.huanchengfly.tieba.post.api.booleanToString
import com.huanchengfly.tieba.post.api.models.CommonResponse
import com.huanchengfly.tieba.post.api.models.ThreadStoreBean
import com.huanchengfly.tieba.post.api.models.ThreadStoreBean.ThreadStoreInfo
import com.huanchengfly.tieba.post.components.MyLinearLayoutManager
import com.huanchengfly.tieba.post.components.MyViewHolder
import com.huanchengfly.tieba.post.dataStore
import com.huanchengfly.tieba.post.getBoolean
import com.huanchengfly.tieba.post.utils.AccountUtil.getLoginInfo
import com.huanchengfly.tieba.post.utils.NavigationHelper
import com.huanchengfly.tieba.post.utils.ThemeUtil.setThemeForSmartRefreshLayout
import com.huanchengfly.tieba.post.utils.Util
import com.scwang.smart.refresh.layout.SmartRefreshLayout
import com.scwang.smart.refresh.layout.api.RefreshLayout
import com.scwang.smart.refresh.layout.listener.OnRefreshLoadMoreListener
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ThreadStoreFragment : BaseFragment() {
    @JvmField
    @BindView(R.id.thread_store_recycler_view)
    var recyclerView: RecyclerView? = null

    @JvmField
    @BindView(R.id.thread_store_refresh_layout)
    var refreshLayout: SmartRefreshLayout? = null
    val navigationHelper: NavigationHelper by lazy { NavigationHelper.newInstance(attachContext) }
    private val threadStoreAdapter: ThreadStoreAdapter by lazy { ThreadStoreAdapter(attachContext) }
    private var page = 0
    private var hasMore = true
    private var tbs: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val account = getLoginInfo()
        if (account != null) tbs = account.tbs
        threadStoreAdapter.setOnItemClickListener(object : OnItemClickListener<ThreadStoreInfo> {
            override fun onClick(viewHolder: MyViewHolder, item: ThreadStoreInfo, position: Int) {
                val map: MutableMap<String, String> = HashMap()
                map["tid"] = item.threadId
                map["pid"] = item.markPid
                map["seeLz"] = attachContext.dataStore.getBoolean("collect_thread_see_lz", true)
                    .booleanToString()
                map["from"] = "collect"
                map["max_pid"] = item.maxPid
                navigationHelper.navigationByData(NavigationHelper.ACTION_THREAD, map)
            }
        })
    }

    override fun getLayoutId(): Int {
        return R.layout.fragment_thread_store
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setThemeForSmartRefreshLayout(refreshLayout!!)
        refreshLayout!!.setOnRefreshLoadMoreListener(object : OnRefreshLoadMoreListener {
            override fun onLoadMore(refreshLayout: RefreshLayout) {
                loadMore()
            }

            override fun onRefresh(refreshLayout: RefreshLayout) {
                refresh()
            }
        })
        recyclerView!!.layoutManager = MyLinearLayoutManager(attachContext)
        val mItemTouchHelper = ItemTouchHelper(object : ItemTouchHelper.Callback() {
            override fun clearView(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder
            ) {
                super.clearView(recyclerView, viewHolder)
                viewHolder.itemView.setBackgroundColor(Color.TRANSPARENT)
            }

            override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
                super.onSelectedChanged(viewHolder, actionState)
                if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
                    viewHolder!!.itemView.setBackgroundColor(
                        Util.getColorByAttr(
                            attachContext,
                            R.attr.colorControlHighlight,
                            R.color.transparent
                        )
                    )
                }
            }

            override fun getMovementFlags(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder
            ): Int {
                val dragFlags = 0
                val swiped = ItemTouchHelper.RIGHT or ItemTouchHelper.LEFT
                return makeMovementFlags(dragFlags, swiped)
            }

            override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder
            ): Boolean {
                return true
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val position = viewHolder.adapterPosition
                val threadStoreInfo = threadStoreAdapter.getItem(position)
                threadStoreAdapter.remove(position)
                Util.createSnackbar(recyclerView!!, R.string.toast_deleted, Snackbar.LENGTH_LONG)
                    .addCallback(object : Snackbar.Callback() {
                        override fun onDismissed(snackbar: Snackbar, event: Int) {
                            if (event != DISMISS_EVENT_ACTION) {
                                getInstance()
                                    .removeStore(threadStoreInfo.threadId, tbs!!)
                                    .enqueue(object : Callback<CommonResponse?> {
                                        override fun onResponse(
                                            call: Call<CommonResponse?>,
                                            response: Response<CommonResponse?>
                                        ) {
                                        }

                                        override fun onFailure(
                                            call: Call<CommonResponse?>,
                                            t: Throwable
                                        ) {
                                            Toast.makeText(
                                                attachContext,
                                                attachContext.getString(
                                                    R.string.toast_delete_error,
                                                    t.message
                                                ),
                                                Toast.LENGTH_SHORT
                                            ).show()
                                            threadStoreAdapter.insert(threadStoreInfo, position)
                                        }
                                    })
                            }
                        }
                    }).setAction(R.string.button_undo) { mView: View? ->
                        threadStoreAdapter.insert(
                            threadStoreInfo,
                            position
                        )
                    }
                    .show()
            }
        })
        mItemTouchHelper.attachToRecyclerView(recyclerView)
        recyclerView!!.adapter = threadStoreAdapter
    }

    override fun onFragmentFirstVisible() {
        refreshLayout!!.autoRefresh()
    }

    private fun refresh() {
        page = 0
        getInstance()
            .threadStore(page, 20)
            .enqueue(object : Callback<ThreadStoreBean?> {
                override fun onResponse(
                    call: Call<ThreadStoreBean?>,
                    response: Response<ThreadStoreBean?>
                ) {
                    val data = response.body()
                    refreshLayout!!.finishRefresh()
                    refreshLayout!!.setNoMoreData(!hasMore)
                    val storeInfoList = data!!.storeThread ?: return
                    threadStoreAdapter.reset()
                    threadStoreAdapter.setData(storeInfoList)
                    hasMore = storeInfoList.size > 0
                }

                override fun onFailure(call: Call<ThreadStoreBean?>, t: Throwable) {
                    refreshLayout!!.finishRefresh(false)
                }
            })
    }

    private fun loadMore() {
        if (!hasMore) {
            return
        }
        getInstance()
            .threadStore(page + 1, 20)
            .enqueue(object : Callback<ThreadStoreBean?> {
                override fun onResponse(
                    call: Call<ThreadStoreBean?>,
                    response: Response<ThreadStoreBean?>
                ) {
                    page += 1
                    val data = response.body()
                    val storeInfoList = data!!.storeThread ?: return
                    threadStoreAdapter.insert(storeInfoList)
                    refreshLayout!!.finishLoadMore()
                    refreshLayout!!.setNoMoreData(!hasMore)
                    hasMore = storeInfoList.size > 0
                }

                override fun onFailure(call: Call<ThreadStoreBean?>, t: Throwable) {
                    refreshLayout!!.finishLoadMore(false)
                    Toast.makeText(attachContext, t.message, Toast.LENGTH_SHORT).show()
                }
            })
    }
}